File: frontend/lib/models/application_event.dart
/// A first-class recruitment event stored in the application_events collection.
///
/// Independent from the embedded [TimelineEvent] list inside [Application].
/// Sorted by [eventDate] (ascending) when displayed.
class ApplicationEvent {
  final String id;
  final String applicationId;
  final String eventType;

  /// ISO date string "YYYY-MM-DD" — may be a future date for scheduled events.
  final String eventDate;

  final String? notes;

  /// ISO-8601 datetime string set server-side at creation. Never sent by client.
  final String createdAt;

  const ApplicationEvent({
    required this.id,
    required this.applicationId,
    required this.eventType,
    required this.eventDate,
    this.notes,
    required this.createdAt,
  });

  factory ApplicationEvent.fromJson(Map<String, dynamic> json) {
    return ApplicationEvent(
      id: json['id'] as String? ?? '',
      applicationId: json['applicationId'] as String? ?? '',
      eventType: json['eventType'] as String? ?? '',
      eventDate: json['eventDate'] as String? ?? '',
      notes: json['notes'] as String?,
      createdAt: json['createdAt'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'applicationId': applicationId,
        'eventType': eventType,
        'eventDate': eventDate,
        'notes': notes,
        'createdAt': createdAt,
      };

  ApplicationEvent copyWith({
    String? id,
    String? applicationId,
    String? eventType,
    String? eventDate,
    String? notes,
    String? createdAt,
  }) {
    return ApplicationEvent(
      id: id ?? this.id,
      applicationId: applicationId ?? this.applicationId,
      eventType: eventType ?? this.eventType,
      eventDate: eventDate ?? this.eventDate,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  /// Parses [eventDate] ("YYYY-MM-DD") into a [DateTime].
  /// Returns null if the string is empty or malformed.
  DateTime? get eventDateTime {
    if (eventDate.isEmpty) return null;
    try {
      return DateTime.parse(eventDate);
    } catch (_) {
      return null;
    }
  }
}

File: backend/demo/src/main/java/com/jobtracker/demo/model/ApplicationEvent.java
package com.jobtracker.demo.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.index.Indexed;

import java.util.List;

/**
 * A first-class recruitment event belonging to one JobApplication.
 * Stored in the "application_events" collection — completely separate
 * from the embedded timeline array inside JobApplication.
 */
@Document(collection = "application_events")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ApplicationEvent {

    /** Valid eventType values. */
    public static final List<String> VALID_EVENT_TYPES = List.of(
            "Applied",
            "OA Scheduled",
            "OA Completed",
            "Interview Scheduled",
            "Interview Completed",
            "HR Round Scheduled",
            "HR Round Completed",
            "Offer Received",
            "Offer Accepted",
            "Rejected",
            "Joined"
    );

    @Id
    private String id;

    /** FK → JobApplication.id */
    @Indexed
    private String applicationId;

    /**
     * Category of recruitment milestone.
     * Must be one of VALID_EVENT_TYPES.
     */
    private String eventType;

    /**
     * Real-world date the event occurred or is scheduled, e.g. "2026-06-20".
     * ISO date string "YYYY-MM-DD". May be in the future for scheduled events.
     */
    private String eventDate;

    /** Optional free-text notes. Null when not provided. */
    private String notes;

    /**
     * Server-set ISO-8601 datetime string, e.g. "2026-06-20T14:30:00".
     * Never supplied by the client — always set by ApplicationEventServiceImpl.
     */
    private String createdAt;
}
